plugins/janus_nosip.c

```diff
@@ -0,0 +1,2213 @@
+	socklen_t len;
+	int fd = socket(AF_INET, SOCK_DGRAM, 0);
+	if (fd == -1)
+		goto error;
+	addr.sin_family = AF_INET;
+	addr.sin_port = htons(1);
+	inet_pton(AF_INET, "1.2.3.4", &addr.sin_addr.s_addr);
+	if (connect(fd, (const struct sockaddr*) &addr, sizeof(addr)) < 0)
+		goto error;
+	len = sizeof(addr);
+	if (getsockname(fd, (struct sockaddr*) &addr, &len) < 0)
+		goto error;
+	if (getnameinfo((const struct sockaddr*) &addr, sizeof(addr),
+			buf, buflen,
+			NULL, 0, NI_NUMERICHOST) != 0)
+		goto error;
+	close(fd);
+	return;
+
+error:
+	if (fd != -1)
+		close(fd);
+	JANUS_LOG(LOG_VERB, "Couldn't find any address! using 127.0.0.1 as the local IP... (which is NOT going to work out of your machine)\n");
+	g_strlcpy(buf, "127.0.0.1", buflen);
+}
+
+
+/* Plugin implementation */
+int janus_nosip_init(janus_callbacks *callback, const char *config_path) {
+	if(g_atomic_int_get(&stopping)) {
+		/* Still stopping from before */
+		return -1;
+	}
+	if(callback == NULL || config_path == NULL) {
+		/* Invalid arguments */
+		return -1;
+	}
+
+	/* Read configuration */
+	char filename[255];
+	g_snprintf(filename, 255, "%s/%s.cfg", config_path, JANUS_NOSIP_PACKAGE);
+	JANUS_LOG(LOG_VERB, "Configuration file: %s\n", filename);
+	janus_config *config = janus_config_parse(filename);
+	if(config != NULL) {
+		janus_config_print(config);
+
+		gboolean local_ip_set = FALSE;
+		janus_config_item *item = janus_config_get_item_drilldown(config, "general", "local_ip");
+		if(item && item->value) {
+			int family;
+			if (!janus_is_ip_valid(item->value, &family)) {
+				JANUS_LOG(LOG_WARN, "Invalid local IP specified: %s, guessing the default...\n", item->value);
+			} else {
+				/* Verify that we can actually bind to that address */
+				int fd = socket(family, SOCK_DGRAM, 0);
+				if (fd == -1) {
+					JANUS_LOG(LOG_WARN, "Error creating test socket, falling back to detecting IP address...\n");
+				} else {
+					int r;
+					struct sockaddr_storage ss;
+					socklen_t addrlen;
+					memset(&ss, 0, sizeof(ss));
+					if (family == AF_INET) {
+						struct sockaddr_in *addr4 = (struct sockaddr_in*)&ss;
+						addr4->sin_family = AF_INET;
+						addr4->sin_port = 0;
+						inet_pton(AF_INET, item->value, &(addr4->sin_addr.s_addr));
+						addrlen = sizeof(struct sockaddr_in);
+					} else {
+						struct sockaddr_in6 *addr6 = (struct sockaddr_in6*)&ss;
+						addr6->sin6_family = AF_INET6;
+						addr6->sin6_port = 0;
+						inet_pton(AF_INET6, item->value, &(addr6->sin6_addr.s6_addr));
+						addrlen = sizeof(struct sockaddr_in6);
+					}
+					r = bind(fd, (const struct sockaddr*)&ss, addrlen);
+					close(fd);
+					if (r < 0) {
+						JANUS_LOG(LOG_WARN, "Error setting local IP address to %s, falling back to detecting IP address...\n", item->value);
+					} else {
+						g_strlcpy(local_ip, item->value, sizeof(local_ip));
+						local_ip_set = TRUE;
+					}
+				}
+			}
+		}
+		if (!local_ip_set)
+			janus_nosip_detect_local_ip(local_ip, sizeof(local_ip));
```

Comment:
Ack and agree! I always try to do that but there may be places where I still forget to do that.